File: movil/test/voz/parser_voz_corpus_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:forja_movil/voz/parser_voz.dart';

import 'corpus_de_voz.dart';

/// La gramatica del telefono contra el MISMO corpus que la del servidor.
///
/// Esta es la prueba que sostiene la promesa de que las dos copias dicen lo
/// mismo. Si se pone roja y ParserVozCorpusTest sigue verde, el telefono se
/// separo; si se ponen rojas las dos, cambio el corpus a proposito.
void main() {
  for (final caso in cargarCorpus()) {
    for (final frase in caso.frases) {
      test('${caso.id}: "$frase"', () {
        verificarCaso(caso, frase, ParserVoz().interpretar(frase, caso.contexto));
      });
    }
  }
}

File: src/test/java/bo/forja/backend/voz/ParserVozCorpusTest.java
package bo.forja.backend.voz;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import java.util.ArrayList;
import java.util.List;

/**
 * La gramatica contra el corpus compartido.
 * <p>
 * Estos casos no viven aca sino en {@code compartido/corpus-voz.json}, porque
 * los lee tambien la suite de Dart del cliente movil. Es lo unico que impide
 * que la copia del telefono y la del servidor se separen sin que nadie se
 * entere: tocar una regla de un lado pone roja la prueba del otro.
 */
@DisplayName("Dictado por voz contra el corpus compartido")
class ParserVozCorpusTest {

    private final ParserVoz parser = new ParserVoz();

    @TestFactory
    List<DynamicTest> elCorpusEntero() {
        List<DynamicTest> pruebas = new ArrayList<>();
        for (CorpusDeVoz.Caso caso : CorpusDeVoz.cargar()) {
            for (String frase : caso.frases()) {
                pruebas.add(DynamicTest.dynamicTest(
                        caso.id() + ": " + frase,
                        () -> CorpusDeVoz.verificar(caso, frase,
                                parser.interpretar(frase, caso.contextoResuelto()))));
            }
        }
        return pruebas;
    }
}
